### PR TITLE
feat(Client/Pool): Same origin redirect

### LINKED ISSUE #686: Add same origin redirect support for Client and Pool
See how it's done in `Agent`. If redirect then create a `RedirectHandler` wrapper around the original dispatch handler.

### PR DESCRIPTION
**Pull request by MetCoder95**
**WIP**

Closes #686 
This PR introduces support for redirect by using `RedirectHandler` on `Client` for support redirection.

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -256,7 +264,10 @@ class Client extends Dispatcher {
     }
 
     try {
-      const request = new Request(opts, handler)
+      console.log('Executing request')
+      const request = this[kMaxRedirections] > 0
+        ? new Request(opts, new RedirectHandler(this, { ...opts, origin: this[kUrl] }, handler))
+        : new Request(opts, handler)
```
Comment (MetCoder95) [inline lib/client.js:270]: cc: @ronag 
I would like to know if I'm following the correct path as when trying to test it, I'm falling into the following error:

```text
error; ClientClosedError: The client is closed
    at Client.dispatch (/Users/carlosgerardofuentesrios/Documents/Workspace/MetCoder95/undici/lib/client.js:277:15)
    at RedirectHandler.onComplete (/Users/carlosgerardofuentesrios/Documents/Workspace/MetCoder95/undici/lib/handler/redirect.js:101:23)
    at Request.onComplete (/Users/carlosgerardofuentesrios/Documents/Workspace/MetCoder95/undici/lib/core/request.js:159:27)
    at Parser.onMessageComplete (/Users/carlosgerardofuentesrios/Documents/Workspace/MetCoder95/undici/lib/client.js:837:15)
    at wasm_on_message_complete (/Users/carlosgerardofuentesrios/Documents/Workspace/MetCoder95/undici/lib/client.js:430:28)
    at null.<anonymous> (wasm://wasm/0002924a:0:1998)
    at null.<anonymous> (wasm://wasm/0002924a:0:6386)
    at null.<anonymous> (wasm://wasm/0002924a:0:3006)
    at llhttp_execute (wasm://wasm/0002924a:0:1055)
    at Parser.execute (/Users/carlosgerardofuentesrios/Documents/Workspace/MetCoder95/undici/lib/client.js:528:32) {
    code: 'UND_ERR_CLOSED'
}
```

File: lib/client.js [hunk 2]
```diff
@@ -256,7 +264,10 @@ class Client extends Dispatcher {
     }
 
     try {
-      const request = new Request(opts, handler)
+      console.log('Executing request')
+      const request = this[kMaxRedirections] > 0
+        ? new Request(opts, new RedirectHandler(this, { ...opts, origin: this[kUrl] }, handler))
```
Comment (MetCoder95) [inline lib/client.js:269]: I also have to inject `origin` based on the sanitized `url` in order to construct the `request.history` properly.

**Comment by MetCoder95** [comment]
Also, I'm not 100% sure, but I think that adding support for `maxRedirections` on the `Client` is enough to propagate the support for the `Pool` object. This due to the fact that the `Pool` creates `Client` instances when building the pool.

File: lib/client.js (hunk 1 above)
Comment (MetCoder95) [reply lib/client.js:270]: Any help? 😅 
I'm running out of options, not sure what's causing the client to be `closed` when the `RedirectHandler` tries to set dispatch the new request. Following the stack trace I only see the happy path once the request is done, but ran out of ideas what's closing the client

Comment (MetCoder95) [reply lib/client.js:249]: Found that the error is mostly in the testing part, I'll continue with it soon 👍

File: lib/client.js [hunk 3]
```diff
@@ -234,7 +243,10 @@ class Client extends Dispatcher {
     }
 
     try {
-      const request = new Request(opts, handler)
+      console.log('Executing request')
```
Comment (mcollina) [inline lib/client.js:246]: I think this should be removed

File: test/client-dispatch.js [hunk 4]
```diff
@@ -124,6 +124,69 @@ test('basic dispatch get', (t) => {
...
+        console.log(buf.toString('utf-8'))
+        bufs.push(buf)
+      },
+      onComplete (trailers) {
+        t.equal(trailers, null)
+        t.equal('hello world!', Buffer.concat(bufs).toString('utf8'))
+      },
+      onError (err) {
+        console.log('error;', err)
+        t.fail()
```
Comment (mcollina) [inline test/client-dispatch.js:184]: replace this with `t.error(err)` and remove the console.log

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/client.js (hunk 2 above)
Comment (ronag) [reply lib/client.js:246]: I'm not sure I follow this.

**Review by ronag — APPROVED** [review]

**Comment by ronag** [comment]
@mcollina @MetCoder95 I finished up this PR. PTAL.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by MetCoder95** [comment]
LGTM, thanks for taking care of the missing parts! 🙂

**Pull request merged**